packages/config-helpers/src/define-config.js

```diff
@@ -0,0 +1,464 @@
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+const eslintrcKeys = [
+	"env",
+	"extends",
+	"globals",
+	"ignorePatterns",
+	"noInlineConfig",
+	"overrides",
+	"parser",
+	"parserOptions",
+	"reportUnusedDisableDirectives",
+	"root",
+];
+
+const allowedGlobalIgnoreKeys = new Set(["ignores", "name"]);
+
+/**
+ * Gets the name of a config object.
+ * @param {Config} config The config object.
+ * @param {string} indexPath The index path of the config object.
+ * @return {string} The name of the config object.
+ */
+function getConfigName(config, indexPath) {
+	if (config.name) {
+		return config.name;
+	}
+
+	return `UserConfig${indexPath}`;
+}
+
+/**
+ * Gets the name of an extension.
+ * @param {SimpleExtendsElement} extension The extension.
+ * @param {string} indexPath The index of the extension.
+ * @return {string} The name of the extension.
+ */
+function getExtensionName(extension, indexPath) {
+	if (typeof extension === "string") {
+		return extension;
+	}
+
+	if (extension.name) {
+		return extension.name;
+	}
+
+	return `ExtendedConfig${indexPath}`;
+}
+
+/**
+ * Determines if a config object is a legacy config.
+ * @param {Config|LegacyConfig} config The config object to check.
+ * @return {config is LegacyConfig} `true` if the config object is a legacy config.
+ */
+function isLegacyConfig(config) {
+	for (const key of eslintrcKeys) {
+		if (key in config) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Determines if a config object is a global ignores config.
+ * @param {Config} config The config object to check.
+ * @return {boolean} `true` if the config object is a global ignores config.
+ */
+function isGlobalIgnores(config) {
+	return Object.keys(config).every(key => allowedGlobalIgnoreKeys.has(key));
+}
+
+/**
+ * Parses a plugin member ID (rule, processor, etc.) and returns
+ * the namespace and member name.
+ * @param {string} id The ID to parse.
+ * @returns {{namespace:string, name:string}} The namespace and member name.
+ */
+function getPluginMember(id) {
+	const firstSlashIndex = id.indexOf("/");
+
+	if (firstSlashIndex === -1) {
+		return { namespace: "", name: id };
+	}
+
+	let namespace = id.slice(0, firstSlashIndex);
+
+	/*
+	 * Special cases:
+	 * 1. The namespace is `@`, that means it's referring to the
+	 *    core plugin so `@` is the full namespace.
+	 * 2. The namespace starts with `@`, that means it's referring to
+	 *    an npm scoped package. That means the namespace is the scope
+	 *    and the package name (i.e., `@eslint/core`).
+	 */
+	if (namespace[0] === "@" && namespace !== "@") {
+		const secondSlashIndex = id.indexOf("/", firstSlashIndex + 1);
+		if (secondSlashIndex === -1) {
+			return { namespace, name: id };
+		}
+		namespace = id.slice(0, secondSlashIndex);
+		return { namespace, name: id.slice(secondSlashIndex + 1) };
```

Comment:
```suggestion
		if (secondSlashIndex !== -1) {
			namespace = id.slice(0, secondSlashIndex);
			return { namespace, name: id.slice(secondSlashIndex + 1) };
		}
```

For example, `"@typescript-eslint/recommended"` should be parsed as `{ namespace: "@typescript-eslint", name: "recommended" }`. This currently parses it as `{ namespace: "@typescript-eslint", name: "@typescript-eslint/recommended" }` (line 120 returns `name: id`).

Reply:
Just a note: `@typescript-eslint/recommended` isn't a valid config name. Plugin names can't be an identifier with a leading `@` unless there is also a slash in the name (like `@eslint/js`) because it needs to mimic scoped packages in npm for backwards compatibility.